_posts/2025-03-18-movie-similarity-search-using-vector-databases.adoc

```diff
@@ -0,0 +1,416 @@
+---
+layout: post
+title: 'Movie similarity search using vector databases'
+date: 2025-03-18
+tags: langchain4j, llm, ai
+synopsis: Shows how to create a movie similarity search system using a vector database and Quarkus Langchain4j
+author: iocanel
+---
+:imagesdir: /assets/images/posts/similarity-search-using-vector-dbs
+ifdef::env-github,env-browser,env-vscode[:imagesdir: ../assets/images/posts/similarity-search-using-vector-dbs]
+
+=== Introduction ===
+With LLMs becoming increasingly popular we often see them being used even for tasks that are not directly related to text generation. 
+Such case is using LLMs for recommendation systems. In this post we'll see how you can build such a system using https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[Quarkus Langchain4j]
+but without using of LLMs. More specifically we'll create a simple movie similarity search system using a vector database. The role
+of https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[Quarkus Langchain4j] in this store is to abstract the underlying vector database through the `EmbeddingStore` interface.
+
+A relevant sample has been recently added to the https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/[Quarkus Langchain4j samples].
+
+=== Embeddings ===
+An embedding is a way to represent unstructured data (e.g. text, images etc) in a structured way. This is done by mapping the data into a vector.
+As we are able to perform mathematical operations on vectors, such as calculating the distance between them, we can use embeddings to calculate 
+how close (or similar) two pieces of data are. In our case, comparing the embeddings of the movie overview can give us a measure of how similar
+the two movies are. This is the premise of this post.
+
+
+==== How are embeddings created? ====
+There are multiple ways to create embeddings. https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[Quarkus Langchain4j] provides the following:
+
+* https://docs.quarkiverse.io/quarkus-langchain4j/dev/in-process-embedding.html[In-process embedding]
+* https://docs.quarkiverse.io/quarkus-langchain4j/dev/llms.html[Model provider]
+
+In this post we'll use the former. Did I mention that we won't be using LLMs?
+Instead, we'll be using https://github.com/pgvector/pgvector[pgvector] which is a PostgreSQL extension that provides vector operations and indexing.
+
+Our project will need the following dependencies, for https://github.com/pgvector/pgvector[pgvector] and https://docs.quarkiverse.io/quarkus-langchain4j/dev/in-process-embedding.html[in-process embedding]:
+
+[source,xml]
+----
+    <dependency>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-embeddings-bge-small-en-q</artifactId>
+        <version>1.0.0-beta1</version>
+    </dependency>
```

Comment:
It might be better to use the Quarkus LangChain4j BOM so that we don't have to define the matching version of LangChain4j